Step5

LarryJung opened the pull request:
step5구현하였습니다. 
추상 클래스 관련해서는 저번 푸쉬 할 때 최대한 고민해보고 한 것인데, 어떠한 부분을 더 수정하면 좋을 지 피드백을 받고 싶습니다!
오전 강의 듣고 positive객체를 추가하면서 포인트간 길이 구할 때 positive에서부터 역할을 위임하는 식으로 구현해봤습니다.

LarryJung committed:
Car exercise add

LarryJung committed:
Positive class add

LarryJung committed:
remove useless method - getX getY in Points class

LarryJung committed:
point, positive class equals hashcode re-overriding, test code add

javajigi commented on src/main/java/car/example/Avante.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    String getName(){
+        return name;
```
굳이 super로 전달할 필요가 있나?
이 메소드에서 "Avante"로 반환한다면?

javajigi commented on src/main/java/car/example/Avante.java:
```diff
@@ -0,0 +1,23 @@
+    public Avante(int tripDistance){
+        super.name = "Avante";
+        super.tripDistance = tripDistance;
+        super.distancePerLiter = 15;
```
super로 값을 전달하지 않고 구현해 본다.

javajigi commented on src/main/java/car/example/Car.java:
```diff
@@ -0,0 +1,19 @@
+
+    double distancePerLiter;
+    String name;
+    double tripDistance;
```
이 3개의 인스턴스 변수를 제거하고 구현해 본다.

javajigi commented on src/test/java/domain/PositiveTest.java:
```diff
@@ -0,0 +1,16 @@
+        Positive positive2 = new Positive(5);
+        assertThat(positive1.getDiff(positive2), is(2));
+    }
+}
```
concreteFigureMaker 클래스 이름이 소문자로 시작함.

javajigi commented on src/test/java/domain/PositiveTest.java:
```diff
@@ -0,0 +1,16 @@
+        Positive positive2 = new Positive(5);
+        assertThat(positive1.getDiff(positive2), is(2));
+    }
+}
```
Figure 객체의 points와 area 인스턴스 변수를 private으로 변경하고 구현 방법을 찾아본다.
앞의 Car 예제를 해결하는 방법과 비슷한 방법으로 구현해 본다.

LarryJung replied:
질문 있습니다.

추상 클래스 안에 package-private으로 선언되어 있는 필드를 private으로 바꾸면 서브 클래스가 넘겨받지 못합니다. 그래서 서브 클래스 안에서 private으로 선언하게 되면, 서브클래스에 각자 써 주어야 하는 중복이 발생하고.. 구현부가 있는 메소드도 사용하지 못하게 되므로 서브 클래스에서 구현해주어야 하는, 인터페이스와 비슷하게 돼버린 것 같은데요.
이게 필드 선언의 안정성의 이유로 이런 식으로 구현하라는 의도인지, 아니면 아예 Figure안에서 private으로 선언 후 구현할 수 있는 방법을 찾아보라는 의도인지 궁금합니다.ㅎㅎ

javajigi reviewed (changes requested):
구현 잘 했네. 피드백 남겨본다.

LarryJung committed:
step5 fix
1. remove package-private field in abstract class, declare these field at each sub-classes with private field. but although this case might be safe code, it seems be duplicated unwillingly.
2. changed figure class like interface. i.e. there are no any instace. be just abstract method.
3. implemented method (not abstract) in figure class was changed abstract. and these methods were implemented at each sub-classes.

LarryJung committed:
fix Car example, remove getName method, override toString()

LarryJung committed:
add parameterized constructor in Car abstract class so that when initializing sub-class private super class's paramters can tbe set by super() constructor. and can be distinguished with sub-class's field / super-classed field

LarryJung committed:
move private Points field in sub-classes to Figure abstract class so that instantiated when initializing
1. some problems happened. initializing exception part in sub-class's constructor was moved to factory construct method 'of' and this is right changing?
2. cause each area calculated method refer to Points instance which is in abstract figure class, should be moved to abstract method from sub-classes.

LarryJung commented:
1. 수퍼생성자 super()는 생성자의 첫번째 라인에 있어야 한다는 것 때문에, 각 서브클래스의 생성자에 추가한 예외처리 부분을 팩토리 메소드 안으로 옮겼습니다. 이와 같이 예외처리를 팩토리생성자 메소드 안에서 해결할 때의 문제점은 없는지?
2. Points 객체를 Figure abstract class 안으로 옮겼기 때문에 각 서브클래스의 넓이를 구하는 식이 abstract class의 구현부로 들어갔습니다. 이 부분이 조금 찜찜한데, 이렇게 될 수 밖에 없는지 아니면 다른 해결방안이 있을지 궁금합니다.

javajigi commented on src/main/java/domain/Figure.java:
```diff
@@ -1,11 +1,45 @@
+    }
+
+    double calculateLine() {
+        return points.calculateLineBwPoints(0, 1);
```
calculateTriangle, calculateRectangle, calculateLine 메소드는 Figure를 상속하는 자식 클래스에서 구현하는 것이 맞지 않을까?
그 로직을 구현하는 곳이 calculate() 추상 메소드를 구현하는 메소드여야 할 것 같은데..

LarryJung replied:
Points써서 길이구하는 메소드가 중복이란 걸 알았고, 다시 리팩토링 해 보았습니다.

javajigi reviewed (changes requested):
마지막 피드백이 되기를..

LarryJung committed:
remove some redundant method

LarryJung committed:
abstract class fix
